### DISCUSSION TITLE
Many-to-many relation and INNER JOIN

### DISCUSSION THREAD

**Discussion by anton-johansson** [prisma/prisma#15873] [Q&A]
## Problem

Imagine a model like this:

```
model User {
    userId                   Int                 @id @default(autoincrement())
    emailAddress             String              @unique @db.NVarChar(100)
    instances                UserInstance[]
}

model Instance {
    instanceId               Int                 @id @default(autoincrement())
    instanceName             String              @db.NVarChar(100)
    users                    UserInstance[]
}

model UserInstance {
    userId                   Int
    user                     User                @relation(fields: [userId], references: [userId])
    instanceId               Int
    instance                 Instance            @relation(fields: [instanceId], references: [instanceId])
    active                   Boolean             @default(true)

    @@id([userId, instanceId])
}
```

There are multiple users and multiple instances. Users are connected to instances, many-to-many.

Now, imagine that I want to retrieve all users for a given instance and whether or not the users are active within that instance. Using raw SQL, I would write this:

```sql
SELECT  [user].userId
,       [user].emailAddress
,       instance.active
FROM    [User] [user]
INNER JOIN UserInstance instance
        ON      instance.userId = [user].userId
        AND     instance.instanceId = 12345
ORDER BY [user].emailAddress
```

This would give me filter on `UserInstance` while allowing me to extract the data from `UserInstance` at the same time.

This seems to be a little tedious with Prisma (or maybe I'm missing something). I've tried things like this:

```typescript

const users = await client.user.findMany({
    select: {
        userId: true,
        emailAddress: true,
        instances: {
            where: {
                instanceId: 12345,
            },
        },
    },
    where: {
        instances: {
            some: {
                instanceId: 12345,
            },
        },
    },
    orderBy: {
        emailAddress: "asc",
    },
});
```

But it's quite inconvenient, because I still get an array of instances even when I know there is exactly one and the SQL generated feels sub-optimal (formatted manually by me):

```sql
SELECT  [dbo].[User].[userId]
,       [dbo].[User].[emailAddress]
FROM    [dbo].[User]
WHERE   [dbo].[User].[userId] IN
(
        SELECT  [t0].[userId]
        FROM    [dbo].[User] AS [t0]
        INNER JOIN [dbo].[UserInstance] AS [j0]
                ON      [j0].[userId] = [t0].[userId]
        WHERE   [j0].[instanceId] = 12345 AND [t0].[userId] IS NOT NULL
)
ORDER BY [dbo].[User].[emailAddress] ASC OFFSET 0 ROWS

SELECT  [dbo].[UserInstance].[userId]
,       [dbo].[UserInstance].[instanceId]
,       [dbo].[UserInstance].[active]
FROM    [dbo].[UserInstance]
WHERE   [dbo].[UserInstance].[instanceId] = 12345 AND [dbo].[UserInstance].[userId] IN ( ... <resulting IDs from above query ... )
```

## Suggested solution

Not sure, really... Ideas?

## Alternatives

Not that I've found.

## Additional context

I'm running with Microsoft SQL server.

**Comment by floelhoeffel** [comment]
Flattening the array should be possible with `result` client extensions as I shared here: https://github.com/prisma/prisma/issues/15074#issuecomment-1282419200

We hope to get them in preview with `4.6.0`

**Reply by floelhoeffel** [reply]
@anton-johansson Client Extensions were shipped with `4.7.0` 

https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions

Please let us know if you have any feedback - thanks!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
